src/transformers/generation_utils.py

```diff
@@ -1918,6 +1924,18 @@ def contrastive_search(
             _, top_k_ids = torch.topk(logit_for_next_step, dim=-1, k=top_k)
             top_k_probs = torch.gather(next_probs, dim=1, index=top_k_ids)
 
+            # Store scores, attentions and hidden_states when required
+            if return_dict_in_generate:
+                if output_scores:
+                    scores += (logit_for_next_step,)
+
+                if output_hidden_states:
+                    decoder_hidden_states += (
+                        (outputs.decoder_hidden_states,)
+                        if self.config.is_encoder_decoder
+                        else (outputs.hidden_states,)
+                    )
```

Comment:
Three changes:
1. Moves the `return_dict_in_generate` block to where it happens in the other generate functions (after getting the logits)
2. Stores the correct hidden states: before it was storing the hidden states for generation step X+1, as opposed to the intended step X
3. Stores the correct scores: I incorrectly suggested to the authors to store the post-contrastive search penalty scores here. Those scores are definitely not logits, so it would break the API with respect to other generation methods. Instead, store the logits here, as in the other functions. (note: we may decide to include the post-contrastive search penalty scores in the future, as part of the output)